extensions/replication/utils/SetupBackbeatClient.js

```diff
@@ -0,0 +1,175 @@
+const http = require('http');
+const errors = require('arsenal').errors;
+const jsutil = require('arsenal').jsutil;
+const VaultClientCache = require('../../../lib/clients/VaultClientCache');
+const BackbeatClient = require('../../../lib/clients/BackbeatClient');
+const BackbeatTask = require('../../../lib/tasks/BackbeatTask');
+const {
+    StaticFileAccountCredentials,
+    ProvisionedServiceAccountCredentials,
+} = require('../../../lib/credentials/AccountCredentials');
+const { attachReqUids } = require('../../../lib/clients/utils');
+const RoleCredentials = require('../../../lib/credentials/RoleCredentials');
+
+class SetupBackbeatClient extends BackbeatTask {
+    constructor(sourceConfig, sourceHTTPAgent) {
+        super();
+        this.sourceConfig = sourceConfig;
+        // TODO: For SSL support, create HTTPS agents instead.
+        this.sourceHTTPAgent = sourceHTTPAgent ||
+            new http.Agent({ keepAlive: true });
+        this._setupVaultclientCache();
+    }
+
+    _setupVaultclientCache() {
+        this.vaultclientCache = new VaultClientCache();
+        if (this.sourceConfig.auth.type === 'role') {
+            const { host, port } = this.sourceConfig.auth.vault;
+            this.vaultclientCache
+                .setHost('source:s3', host)
+                .setPort('source:s3', port);
+        }
+    }
+
+    _createCredentials(log) {
+        const authConfig = this.sourceConfig.auth;
+        if (authConfig.type === 'account') {
+            return new StaticFileAccountCredentials(authConfig, log);
+        }
+        if (authConfig.type === 'service') {
+            return new ProvisionedServiceAccountCredentials(authConfig, log);
+        }
+        const vaultclient = this.vaultclientCache.getClient('source:s3');
+        const extension = 'replication';
+        const role = this.sourceRole;
+        return new RoleCredentials(vaultclient, extension, role, log);
+    }
+
+    putMetadata(entry, log, cb) {
+        this.retry({
+            actionDesc: 'update metadata on source',
+            logFields: { entry: entry.getLogInfo() },
+            actionFunc: done => this._putMetadataOnce(entry, log, done),
+            shouldRetryFunc: err => err.retryable,
+            log,
+        }, cb);
+    }
+
+    _putMetadataOnce(entry, log, cb) {
+        log.debug('putting metadata',
+                  { where: 'source', entry: entry.getLogInfo(),
+                    replicationStatus: entry.getReplicationStatus() });
+
+        // sends extra header x-scal-replication-content to the target
+        // if it's a metadata operation only
+        const mdBlob = entry.getSerialized();
+        const req = this.backbeatSource.putMetadata({
+            Bucket: entry.getBucket(),
```

Comment:
I'm all for factoring the code that sets up the backbeat client and handles GET/PUT of metadata with retries etc., but I would rather not use polymorphism on the `entry` object to support this, as it makes things unnecessarily error-prone (one has to understand and maintain the interface on the `entry` object required to execute the actions, say, for `getBucket()` etc.).

A more direct approach to this is to have the helper method ask for the specific params it needs to do the job (`bucket`, `objectKey` etc.) either through a list of parameters or through a `params` object. Then each user gets and provides its own parameters as he wishes (e.g. by calling `entry.getBucket()` etc) before calling the helper method, which decouples the caller logic from the callee logic.

Reply:
That makes sense. I've updated the `getMetadata` helpers since the `putMetadata` isn't affected in the same way by polymorphism.